### PR TITLE
Update usaco-616.mdx + add java/cpp code

### PR DESCRIPTION
**Pull request by varunragu23**
Java code from official analysis, added both java and cpp code for the cpi video program.

### CODE REVIEW

File: solutions/usaco-616.mdx [hunk 1]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
-	for (int start = 0; start < n; start++) {
-		int cur = 0, cur_cows = cows;
-		for (int i = 0; i < n; i++) {
-			cur_cows -= a[(start + i) % n];
-			cur += cur_cows;
-		}
-		sol = min(sol, cur);
-	}
-	cout << sol << '\n';
+int main() {
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:19]: use tabs instead of spaces to indent

File: solutions/usaco-616.mdx [hunk 2]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+
+  // check each possible barn we can open
+
+  int ret = (int)1e9; // stores the current answer
+  // initially set to an abnormally large number
+
+  for(int open_barn = 0; open_barn < n; open_barn++) {
+    int curdist = 0; // stores the current distance if barn i was the opened barn
+    for(int pos = 0; pos < n; pos++) {
+      // the current barn we are processing is open_barn + pos
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:47]: use multiline comments?

File: solutions/usaco-616.mdx [hunk 3]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+			cows[i] = Integer.parseInt(br.readLine());
+		}
+		// the answer cannot exceed N * N * 100, since there are at most 100N cows, each of which can move at most N
+		int ans = n * n * 100;
+		for(int unlock = 0; unlock < n; unlock++) {
+			// assume we unlock the door at index "unlock", compute the distance all cows travel
+			int currentDistance = 0;
+			for(int offset = 0; offset < n; offset++) {
+				// count how many cows have to walk a distance of "offset"
+				currentDistance += offset * cows[(unlock+offset)%n];
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:92]:
```suggestion
				currentDistance += offset * cows[(unlock + offset) % n];
```

File: solutions/usaco-616.mdx [hunk 4]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+  // initially set to an abnormally large number
+
+  for(int open_barn = 0; open_barn < n; open_barn++) {
+    int curdist = 0; // stores the current distance if barn i was the opened barn
+    for(int pos = 0; pos < n; pos++) {
+      // the current barn we are processing is open_barn + pos
+      // however, since it is a circular array 
+      // the position may wrap around
+      // index = (open_barn + pos)
+      curdist += pos * cows[(open_barn + pos)%n];
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:51]:
```suggestion
      curdist += pos * cows[(open_barn + pos) % n];
```

File: solutions/usaco-616.mdx [hunk 5]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+  int ret = (int)1e9; // stores the current answer
+  // initially set to an abnormally large number
+
+  for(int open_barn = 0; open_barn < n; open_barn++) {
+    int curdist = 0; // stores the current distance if barn i was the opened barn
+    for(int pos = 0; pos < n; pos++) {
+      // the current barn we are processing is open_barn + pos
+      // however, since it is a circular array 
+      // the position may wrap around
+      // index = (open_barn + pos)
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:50]:
```suggestion
      // index = open_barn + pos
```

File: solutions/usaco-616.mdx [hunk 6]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
 ## Implementation
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:13]: you can add your name to the author list at the top

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-616.mdx [hunk 7]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+    cin >> cows[i];
+  }
+
+  // simple brute force
+
+  // check each possible barn we can open
+
+  int ret = (int)1e9; // stores the current answer
+  // initially set to an abnormally large number
+
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:43]:
```suggestion
  int ret = INT32_MAX;
```

File: solutions/usaco-616.mdx [hunk 8]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+  vector<int> cows(n);
+
+  for(int i = 0; i < n; i++) {
+    cin >> cows[i];
+  }
+
+  // simple brute force
+
+  // check each possible barn we can open
+
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:40]:
```suggestion
  // run a simple brute force algo, checking each possible barn we can open

File: solutions/usaco-616.mdx [hunk 9]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+  for(int open_barn = 0; open_barn < n; open_barn++) {
+    int curdist = 0; // stores the current distance if barn i was the opened barn
+    for(int pos = 0; pos < n; pos++) {
+      // the current barn we are processing is open_barn + pos
+      // however, since it is a circular array 
+      // the position may wrap around
+      // index = open_barn + pos
+      curdist += pos * cows[(open_barn + pos) % n];
+    }
+    // update the answer
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:53]:
```suggestion
```

File: solutions/usaco-616.mdx [hunk 10]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+    ret = min(ret, curdist);
+  }
+
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:56]:
```suggestion
```

File: solutions/usaco-616.mdx [hunk 11]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+<JavaSection>
+	
+## Implementation	
+	
+```java
+import java.io.*;
+import java.util.*;
+public class cbarn {
+	public static void main(String[] args) throws IOException {
+		// initialize file I/O
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:75]:
```suggestion
```

File: solutions/usaco-616.mdx [hunk 12]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
 </CPPSection>
+	
+<JavaSection>
+	
+## Implementation	
+	
+```java
+import java.io.*;
+import java.util.*;
+public class cbarn {
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:73]:
```suggestion

public class CBarn {
```

File: solutions/usaco-616.mdx [hunk 13]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
+			// update the answer
+			if(currentDistance < ans) {
+				ans = currentDistance;
+			}
+		}
+		// print the answer
+		pw.println(ans);
+		// close output stream
+		pw.close();
+		br.close();
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:103]: lines 78, 81, 94, 99, and 101 are unnecessary comments (they shouldn't explain what you're doing, but explain _why_ you're doing what you're doing!)

File: solutions/usaco-616.mdx [hunk 14]
```diff
@@ -13,43 +13,98 @@ author: Óscar Garries
...
-	cout << sol << '\n';
+int main() {
+  freopen("cbarn.in", "r", stdin);
+  freopen("cbarn.out", "w", stdout);
+  cin.sync_with_stdio(0); cin.tie(0);
+
+  // input
+
+  // have the number of cows and desired positions of each cow
+
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:27]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-616.mdx [hunk 15]
```diff
@@ -13,43 +13,87 @@ author: Óscar Garries
 ## Implementation
 
 ```cpp
-#include <bits/stdc++.h>
-
+#include<bits/stdc++.h>
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:16]:
```suggestion
#include <bits/stdc++.h>
```

File: solutions/usaco-616.mdx [hunk 16]
```diff
@@ -13,43 +13,87 @@ author: Óscar Garries
...
-		int cur = 0, cur_cows = cows;
-		for (int i = 0; i < n; i++) {
-			cur_cows -= a[(start + i) % n];
-			cur += cur_cows;
+
+	// run a simple brute force algo, checking each possible barn we can open
+
+	int ret = (int)1e9; // stores the current answer
+	// initially set to an abnormally large number
+
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:37]:
```suggestion
	// run a simple brute force algo that checks each possible barn
	int ret = INT32_MAX;
```

File: solutions/usaco-616.mdx [hunk 17]
```diff
@@ -13,43 +13,87 @@ author: Óscar Garries
...
+			curdist += pos * cows[(open_barn + pos) % n];
 		}
-		sol = min(sol, cur);
+		ret = min(ret, curdist);
 	}
-	cout << sol << '\n';
+
+  cout << ret << '\n';
+
+  return 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:53]:
```suggestion
```

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-616.mdx [hunk 18]
```diff
@@ -14,42 +14,86 @@ author: Óscar Garries
...
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));
+		
+		int n = Integer.parseInt(br.readLine());
+		int[] cows = new int[n];
+		for(int i = 0; i < n; i++) {
+			cows[i] = Integer.parseInt(br.readLine());
+		}
+		// the answer cannot exceed N * N * 100, since there are at most 100N cows, each of which can move at most N
```
Comment (brad182) [inline solutions/usaco-616.mdx:77]: this line is over the limit

File: solutions/usaco-616.mdx [hunk 19]
```diff
@@ -14,42 +14,86 @@ author: Óscar Garries
...
+		int ans = n * n * 100;
+		for(int unlock = 0; unlock < n; unlock++) {
+			// assume we unlock the door at index "unlock", compute the distance all cows travel
+			int currentDistance = 0;
+			for(int offset = 0; offset < n; offset++) {
+				// count how many cows have to walk a distance of "offset"
+				currentDistance += offset * cows[(unlock+offset)%n];
+			}
+			if(currentDistance < ans) {
```
Comment (brad182) [inline solutions/usaco-616.mdx:86]: can be simplified using minimum

File: solutions/usaco-616.mdx [hunk 20]
```diff
@@ -14,42 +14,86 @@ author: Óscar Garries
...
+				ans = currentDistance;
+			}
+		}
+		pw.println(ans);
+		pw.close();
```
Comment (brad182) [inline solutions/usaco-616.mdx:91]: i dont think kattio is require but should look nicer its just one close

**Review by brad182 — CHANGES_REQUESTED** [review]

**Comment by varunragu23** [comment]
I just copied the java code from the official analysis, 🤔.

File: solutions/usaco-616.mdx [hunk 21]
```diff
@@ -14,42 +14,117 @@ author: Óscar Garries
...
+			curdist += pos * cows[(open_barn + pos) % n];
 		}
-		sol = min(sol, cur);
+		ret = min(ret, curdist);
 	}
-	cout << sol << '\n';
+
+  cout << ret << '\n';
+
+  return 0;
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:53]: indentation here is weird

File: solutions/usaco-616.mdx [hunk 22]
```diff
@@ -14,42 +14,117 @@ author: Óscar Garries
...
 }
 
 ```
 
 </CPPSection>
+	
+<JavaSection>
+	
+## Implementation
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:62]: move ## Implementation outside languagesection and remove ## Implementation from cppsection

File: solutions/usaco-616.mdx [hunk 23]
```diff
@@ -14,42 +14,117 @@ author: Óscar Garries
...
+	
+## Implementation	
+	
+```java
+import java.io.*;
+import java.util.*;
+public class cbarn {
+	public static void main(String[] args) throws IOException {
+		BufferedReader br = new BufferedReader(new FileReader("cbarn.in"));
+		PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("cbarn.out")));
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:70]: don't need these since you're using kattio, right?

File: solutions/usaco-616.mdx [hunk 24]
```diff
@@ -14,42 +14,117 @@ author: Óscar Garries
...
+				// count how many cows have to walk a distance of "offset"
+				currentDistance += offset * cows[(unlock+offset)%n];
+			}
+			ans = Math.min(ans, currentDistance);
+		}
+		io.println(ans);
+		io.close();
+	}
+	// https://usaco.guide/general/io?lang=java#io-template
+	static class Kattio extends PrintWriter {
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:95]: replace the entire kattio with `CodeSnip{Kattio}`

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

File: solutions/usaco-616.mdx [hunk 25]
```diff
@@ -2,54 +2,94 @@
 id: usaco-616
 source: USACO Bronze 2016 February
 title: Circular Barn
-author: Óscar Garries
+author: Varun Ragunath
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_cbarn_bronze_feb16.html)
 
+## Implementation
```
Comment (maggie-j-liu) [inline solutions/usaco-616.mdx:10]: add time complexity

File: solutions/usaco-616.mdx (hunk 3 above)
Comment (maggie-j-liu) [reply solutions/usaco-616.mdx:82]: bump https://github.com/cpinitiative/usaco-guide/pull/1866/files#r663159957

**Review by maggie-j-liu — CHANGES_REQUESTED** [review]

**Review by maggie-j-liu — APPROVED** [review]
looks good, just address sans's requested changes
https://github.com/cpinitiative/usaco-guide/pull/1866/files#r663189532
https://github.com/cpinitiative/usaco-guide/pull/1866/files#r663189670

File: solutions/usaco-616.mdx [hunk 26]
```diff
@@ -2,54 +2,94 @@
...
+			// index = open_barn + pos
+			curdist += pos * cows[(open_barn + pos) % n];
 		}
-		sol = min(sol, cur);
+		ret = min(ret, curdist);
 	}
-	cout << sol << '\n';
+
+	cout << ret << '\n';
+	return 0;
```
Comment (SansPapyrus683) [inline solutions/usaco-616.mdx:53]:
```suggestion
```

**Pull request closed without merging**